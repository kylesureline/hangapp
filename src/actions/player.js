import database from '../firebase/firebase';

export const initializeGame = (word, guessesRemaining) => ({
  type: 'INIT_GAME',
  answer: word,
  guessesRemaining
});

export const guessLetter = (letter) => ({
  type: 'GUESS_LETTER',
  letter
});

export const updateGuessedWord = (wordArray) => ({
  type: 'UPDATE_GUESSED_WORD',
  wordArray
});

export const wrongGuess = () => ({
  type: 'WRONG_GUESS'
});

export const setPlayer = (player) => ({
  type: 'SET_PLAYER',
  player
});

export const setPlayerStats = (stats) => ({
  type: 'SET_PLAYER_STATS',
  stats
});

export const setPastGames = (pastGames) => ({
  type: 'SET_PAST_GAMES',
  pastGames
});

export const startSetPlayer = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    return database.ref(`players/${uid}`)
      .once('value')
      .then((snapshot) => {
        const pastGames = [];

        for(let game in snapshot.val().pastGames) {
          pastGames.push(snapshot.val().pastGames[game]);
        }

        const stats = snapshot.val().stats;

        dispatch(setPlayerStats(stats));
        dispatch(setPastGames(pastGames));
      });

  };
};
