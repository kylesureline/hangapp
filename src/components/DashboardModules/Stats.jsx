import React from 'react';
import { connect } from 'react-redux';

export const Stats = (props) => (
  <div>
    <h3 className="module__title">Stats</h3>
    <div className="stats__group">
      <p>Wins: {props.stats.wins}</p>
      <p>Losses: {props.stats.losses}</p>
    </div>
  </div>
);

const mapStateToProps = (state) => ({
  stats: state.player.stats
});

export default connect(mapStateToProps)(Stats);
