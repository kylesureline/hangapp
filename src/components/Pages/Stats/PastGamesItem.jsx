import React from 'react';
import { toRelevantTimeString } from '../../../utils';

export const PastGamesItem = ({ game }) => {
  const { answer, timestamp, won } = game;

  return (
    <li className="past-games__item">
      <span>{answer.words.join(' ')}</span>
      <span>{won ? 'won' : 'lost'}</span>
      <span>{toRelevantTimeString(timestamp)}</span>
    </li>
  );
};
