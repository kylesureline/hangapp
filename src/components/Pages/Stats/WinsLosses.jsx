import React from 'react';

export const WinsLosses = ({ data }) => {
  const winCount = data.filter(game => game.won).length;
  const winPerc = (winCount / data.length) * 100;
  const lostCount = data.filter(game => !game.won).length;
  const lostPerc = (lostCount / data.length) * 100;
  return (
    <div className="wins-losses">
      <div
        className="wins"
        style={{
          // flex: `0 1 ${winPerc}%`
          width: `${winPerc}%`,
        }}
        >{winCount} ({winPerc.toFixed(0)}%)
      </div>
      <div
        className="losses"
        style={{
          // flex: `0 1 ${lostPerc}%`
          width: `${lostPerc}%`,
        }}
        >{lostCount} ({lostPerc.toFixed(0)}%)
      </div>
    </div>
  );
};
