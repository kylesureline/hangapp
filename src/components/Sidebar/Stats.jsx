import React from 'react';

export const Stats = () => (
  <ul>
    <li>Wins: <span>0</span></li>
    <li>Losses: <span>0</span></li>
    <li>Cache: <span>0%</span></li>
    <li>Version: <span>1.51</span></li>
  </ul>
);
