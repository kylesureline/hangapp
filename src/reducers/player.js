const defaultGameState = {
  answer: {
    word: '',
    type: '',
    def: ''
  },
  guessedLetters: [],
  guessedWord: [],
  guessesRemaining: 0,
  stats: {
    wins: 0,
    losses: 0
  }
};

export default (state = defaultGameState, action) => {
  switch(action.type) {
    case 'GUESS_LETTER':
      return {
        ...state,
        guessedLetters: [...state.guessedLetters, action.letter]
      };
    case 'UPDATE_GUESSED_WORD':
      return {
        ...state
      };
    case 'WRONG_GUESS':
      return {
        ...state,
        guessesRemaining: state.guessesRemaining -= 1
      };
    case 'SET_PLAYER_STATS':
      return {
        ...state,
        stats: action.stats
      };
    case 'SET_PAST_GAMES':
      return {
        ...state,
        pastGames: action.pastGames
      };
    case 'SET_GUESSES_REMAINING':
      return {
        ...state,
        guessesRemaining: action.guessesRemaining
      }
    case 'CHOOSE_DEFAULT_WORD':
      return {
        ...state,
        answer: {
          word: 'hangapp',
          type: 'noun',
          def: 'A hangman game built as a React PWA by Kyle Scheuerlein'
        }
      };
    case 'CHOOSE_RANDOM_WORD':
      return {
        ...state,
        answer: action.answer
      }
    case 'ADD_WORD':
      return {
        ...state,
        pastGames: [
          ...state.pastGames,
          action.answer
        ]
      }
    case 'ADD_WIN':
      return {
        ...state,
        stats: {
          ...state.stats,
          wins: state.stats.wins + 1
        }
      }
    case 'ADD_LOSS':
      return {
        ...state,
        stats: {
          ...state.stats,
          losses: state.stats.losses + 1
        }
      }
    default:
      return state;
  }
};
