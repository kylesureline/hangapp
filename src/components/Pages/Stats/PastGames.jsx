import React from 'react';
import { PastGamesItem } from './PastGamesItem';

export const PastGames = ({ data }) => {
  return (
    <ul className="past-games">
      {data.map(game => {
        return <PastGamesItem key={game.timestamp} game={game} />;
      })}
    </ul>
  );
};
