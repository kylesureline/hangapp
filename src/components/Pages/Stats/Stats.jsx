import React from "react";
import { PastGames } from "./PastGames";
import { WinsLosses } from "./WinsLosses";
import { getFromLS } from "../../../utils";
import DocumentTitle from "react-document-title";

export const Stats = () => {
  const data = getFromLS("past-games") || [];

  return (
    <DocumentTitle title={`${process.env.REACT_APP_NAME} | Stats`}>
      <main className="page page--stats">
        {!!data.length ? (
          <>
            <div className="stat-widget stat-widget--wins-losses">
              <h2 className="header">wins / losses</h2>
              <WinsLosses data={data} />
            </div>
            <div className="stat-widget stat-widget--past-games">
              <h2 className="header">past games</h2>
              <PastGames data={data} />
            </div>
          </>
        ) : (
          <div className="no-games">no past games to display...</div>
        )}
      </main>
    </DocumentTitle>
  );
};
