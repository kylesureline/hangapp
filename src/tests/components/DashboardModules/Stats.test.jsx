import React from 'react';
import { shallow } from 'enzyme';
import { Stats } from '../../../components/DashboardModules/Stats.jsx';
import { player } from '../../fixtures/player';

test('should render Stats correctly', () => {
  const wrapper = shallow(<Stats stats={player.stats} />);
  expect(wrapper).toMatchSnapshot();
});
