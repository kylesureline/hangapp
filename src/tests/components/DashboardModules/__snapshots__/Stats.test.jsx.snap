// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Stats correctly 1`] = `
<div>
  <h3
    className="module__title"
  >
    Stats
  </h3>
  <div
    className="stats__group"
  >
    <p>
      Wins: 
      5
    </p>
    <p>
      Losses: 
      2
    </p>
  </div>
</div>
`;
